### PR TITLE
{HDInsight} Support creating cluster with ADLSGen1/2 storage in national clouds

### PR DESCRIPTION
**Pull request by aim-for-better**
## Description
Support creating cluster with ADLSGen1/2 storage in national clouds

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs [hunk 1]
```diff
@@ -225,5 +226,21 @@ public virtual void RestartHosts(string resourceGroupName, string clusterName, I
         {
             HdInsightManagementClient.VirtualMachines.RestartHosts(resourceGroupName, clusterName, hosts);
         }
+        
+        private void ResetClusterIdentity(ClusterCreateParametersExtended createParams, string cloudAadAuthority, string cloudDataLakeAudience)
```
Comment (idear1203) [inline src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs:230]: Nit - I would drop the word cloud. E.g `cloudAadAuthority` -> `aadAuthority`

File: src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs [hunk 2]
```diff
@@ -225,5 +226,21 @@ public virtual void RestartHosts(string resourceGroupName, string clusterName, I
...
+            Dictionary<string, string> clusterIdentity;
+            if(!configuation.TryGetValue("clusterIdentity", out clusterIdentity))
+            {
+                return;
+            }
+            clusterIdentity.SetProperty("clusterIdentity.resourceUri", cloudDataLakeAudience);
+
+            string aadTenantIdWithUrl;
+            clusterIdentity.TryGetValue("clusterIdentity.aadTenantId", out aadTenantIdWithUrl);
+            string newAadTenantIdWithUrl = aadTenantIdWithUrl?.Replace("https://login.windows.net/", cloudAadAuthority);
```
Comment (idear1203) [inline src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs:242]: Nit - nice to extract a constant for "https://login.windows.net/"

File: src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs [hunk 3]
```diff
@@ -225,5 +226,21 @@ public virtual void RestartHosts(string resourceGroupName, string clusterName, I
...
+        
+        private void ResetClusterIdentity(ClusterCreateParametersExtended createParams, string cloudAadAuthority, string cloudDataLakeAudience)
+        {
+            var configuation = (Dictionary<string, Dictionary<string, string>>)createParams.Properties.ClusterDefinition.Configurations;
+            Dictionary<string, string> clusterIdentity;
+            if(!configuation.TryGetValue("clusterIdentity", out clusterIdentity))
+            {
+                return;
+            }
+            clusterIdentity.SetProperty("clusterIdentity.resourceUri", cloudDataLakeAudience);
```
Comment (idear1203) [inline src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs:238]: Nit - you can just say `clusterIdentity["clusterIdentity.resourceUri"] = cloudDataLakeAudience;` for simplicity. BTW, what if `clusterIdentity["clusterIdentity.resourceUri"]` is not set originally?

Comment (sunsw1994) [inline src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs:238]: Just throw a question out of curiosity. Is there such a scenario that `clusterIdentity` is not null but `cloudAadAuthority` /`cloudDataLakeAudience` is null?  I am not familiar with  `clusterIdentity`, so if my question is wrong, please correct it.

File: src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs (hunk 2 above)
Comment (sunsw1994) [reply src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs:242]: I agree with Dongwei.

File: src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs (hunk 1 above)
Comment (aim-for-better) [reply src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs]: > Nit - I would drop the word cloud. E.g `cloudAadAuthority` -> `aadAuthority`

Changed.

File: src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs (hunk 2 above)
Comment (aim-for-better) [reply src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs]: > Nit - nice to extract a constant for "https://login.windows.net/"

Fixed.

File: src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs (hunk 3 above)
Comment (aim-for-better) [reply src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs]: > Nit - you can just say `clusterIdentity["clusterIdentity.resourceUri"] = cloudDataLakeAudience;` for simplicity. BTW, what if `clusterIdentity["clusterIdentity.resourceUri"]` is not set originally?

Fixed. If it is not set originally, then our code will add the key/value pair, I think it is ok.

Comment (aim-for-better) [reply src/HDInsight/HDInsight/Models/Management/AzureHdInsightManagementClient.cs]: > Just throw a question out of curiosity. Is there such a scenario that `clusterIdentity` is not null but `cloudAadAuthority` /`cloudDataLakeAudience` is null? I am not familiar with `clusterIdentity`, so if my question is wrong, please correct it.

Yes, `cloudDataLakeAudience` may be null. A dictionary can accept a null as value. From the code side, it is ok.

**Pull request merged**